### PR TITLE
Changes in advance of design update to MatrixDelayUpdateCUDA and DDB

### PR DESCRIPTION
**Pull request by PDoakORNL**
## Proposed changes

API updates for DiracDeterminantComputeCUDA and DiracDeterminantComputeOMPTarget. This is to support changes in classes above to move away from the confused 1 walker to 1 batched object and then a batched leader design which is left over from the 1 manager per walker with only the index 0 being an actual manager design of legacy.  SingleWalker API that are not exposed above DiracDeterminantBatched also have a resource argument added to allow passing in thread local resource handles making it possible to use streams to efficiently transfer and launch kernels without OMP target. An dummy resource needs to be passed by OMPtarget code.

The reason these two classes should be 1 per batch is simple. It makes sense for the batched compute objects to manage or acquire the temporary resources used for used computation, i.e. streams, scratch host or device memory. It doesn't make sense to have 1 of a set of identical 1 per walker classes acquire the compute and memory resource needed for batching and operate on its peers and then just not use those resources for single walker calls.

This is a first step to fixing that from the bottom up.
 
## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- New feature
- Code style update (formatting, renaming)
- Refactoring (no functional changes, no api changes)
- Documentation changes
- Other (please describe):

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
Leconte 

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes This PR is up to date with current the current state of 'develop'
- Yes Code added or changed in the PR has been clang-formatted
- Yes This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by ye-luo** [comment]
Missing DummyResource and please add it to `Utilities/for_testing`

**Comment by ye-luo** [comment]
> Missing DummyResource and please add it to `Utilities/for_testing`

OK. not needed. This is not a separate file.

**Comment by PDoakORNL** [comment]
DummyResource is not just for testing.  The next level uses it.  I'm not giving up yet on native CUDA single walker API's.

File: src/QMCWaveFunctions/Fermion/DiracMatrixComputeOMPTarget.hpp [hunk 1]
```diff
@@ -156,39 +155,44 @@ class DiracMatrixComputeOMPTarget : public Resource
...
-  template<typename TMAT, typename TREAL>
-  inline std::enable_if_t<std::is_same<T_FP, TMAT>::value> invert_transpose(const OffloadPinnedMatrix<TMAT>& a_mat,
-                                                                            OffloadPinnedMatrix<TMAT>& inv_a_mat,
-                                                                            std::complex<TREAL>& log_value)
+  template<typename TMAT>
+  inline std::enable_if_t<std::is_same<VALUE_FP, TMAT>::value> invert_transpose(
+      Resource& resource,
+      const OffloadPinnedMatrix<TMAT>& a_mat,
+      OffloadPinnedMatrix<TMAT>& inv_a_mat,
+      OffloadPinnedVector<std::complex<FullPrecReal>>& log_values)
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/DiracMatrixComputeOMPTarget.hpp:167]: Why is this an array of only 1 element?

File: src/QMCWaveFunctions/Fermion/DiracMatrixComputeCUDA.hpp [hunk 2]
```diff
@@ -297,21 +309,21 @@ class DiracMatrixComputeCUDA : public Resource
...
-      RefVector<OffloadPinnedMatrix<TMAT>>& a_mats,
-      RefVector<OffloadPinnedMatrix<TMAT>>& inv_a_mats,
-      OffloadPinnedVector<std::complex<TREAL>>& log_values,
+      RefVector<DualMatrix<TMAT>>& a_mats,
+      RefVector<DualMatrix<TMAT>>& inv_a_mats,
+      DualVector<std::complex<FullPrecReal>>& log_values,
       const std::vector<bool>& compute_mask)
   {
-    assert(log_values.size() == a_mats.size());
+    //assert(log_values.size() == a_mats.size());
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/DiracMatrixComputeCUDA.hpp:323]: Change?

File: src/QMCWaveFunctions/Fermion/DiracMatrixComputeCUDA.hpp [hunk 3]
```diff
@@ -169,8 +178,8 @@ class DiracMatrixComputeCUDA : public Resource
     LU_diags_fp_.resize(n * nw);
     cudaStream_t hstream;
     cublasErrorCheck(cublasGetStream(h_cublas, &hstream), "cublasGetStream failed!");
-    assert(hstream == hstream_);
-    cudaErrorCheck(cudaMemcpyAsync(psi_Ms.device_data(), psi_Ms.data(), psi_Ms.size() * sizeof(T_FP),
+    //assert(hstream == hstream_);
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracMatrixComputeCUDA.hpp:181]: I'll look over it, the original idea was hstream_ was the stream that data related to this computation was being transferred or sent to kernels on.  A change in that corresponded to a serious logic error.  In the past when I have removed things like this before a feature was actually merged I have regretted it.

File: src/QMCWaveFunctions/Fermion/DiracMatrixComputeCUDA.hpp [hunk 4]
```diff
@@ -221,32 +230,35 @@ class DiracMatrixComputeCUDA : public Resource
    *  vector is used to match the primary batched interface to the accelerated routings.
    *  There is no optimization (yet) for TMAT same type as TREAL
    */
-  template<typename TMAT, typename TREAL>
-  void invert_transpose(CUDALinearAlgebraHandles& cuda_handles,
-                        OffloadPinnedMatrix<TMAT>& a_mat,
-                        OffloadPinnedMatrix<TMAT>& inv_a_mat,
-                        OffloadPinnedVector<std::complex<TREAL>>& log_values)
+  template<typename TMAT>
+  void invert_transpose(Resource& resource,
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracMatrixComputeCUDA.hpp:234]: I didn't decide on this resource framework. This resource comes from the caller and is managed with the ResourceCollection eventually I'd like to have more of the code shared between openMP and CUDA. We can revisit it when the dust settles, I have no interest in rewriting my work in MatrixDelayedUpdateCUDA and DiracBatchedDeterminant again.

File: src/QMCWaveFunctions/Fermion/DiracMatrixComputeCUDA.hpp (hunk 2 above)
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracMatrixComputeCUDA.hpp:323]: yes if there is still a place where this isn't true when interacting with the application that should get fixed.

File: src/QMCWaveFunctions/Fermion/DiracMatrixComputeCUDA.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCWaveFunctions/Fermion/DiracMatrixComputeCUDA.hpp:234]: I thought you pass in explicit type from MatrixDelayedUpdateCUDA already. I doubt change Resource to CUDALinearAlgebraHandles changes anything on the caller side.

[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracMatrixComputeCUDA.hpp:236]: No, for now these two class share a common API.

File: src/QMCWaveFunctions/Fermion/DiracMatrixComputeOMPTarget.hpp [hunk 5]
```diff
@@ -156,39 +155,44 @@ class DiracMatrixComputeOMPTarget : public Resource
...
    * @tparam TMAT matrix value type
    * @tparam TREAL real type
    */
-  template<typename TMAT, typename TREAL>
-  inline std::enable_if_t<std::is_same<T_FP, TMAT>::value> invert_transpose(const OffloadPinnedMatrix<TMAT>& a_mat,
-                                                                            OffloadPinnedMatrix<TMAT>& inv_a_mat,
-                                                                            std::complex<TREAL>& log_value)
+  template<typename TMAT>
+  inline std::enable_if_t<std::is_same<VALUE_FP, TMAT>::value> invert_transpose(
+      Resource& resource,
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracMatrixComputeOMPTarget.hpp:164]: No for now these classes share a common API.

File: src/QMCWaveFunctions/Fermion/DiracMatrixComputeCUDA.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCWaveFunctions/Fermion/DiracMatrixComputeCUDA.hpp:236]: Did you refer to DiracMatrixComputeCUDA.hpp and DiracMatrixComputeOMPTarget.hpp two classes?
One have DualMatrix and the other one have OffloadPinnedMatrix. They are not the same API. You always know what exact type of Resource is.
```
auto& cuda_handles = dynamic_cast<CUDALinearAlgebraHandles&>(resource);
```
this downcast is not necessary and dangerous without checks.

[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracMatrixComputeCUDA.hpp:236]: Its not without a check its a reference, std::bad_cast is thrown.  That's exactly how I'd like it to work.

The types match when both are compiled in the case of ENABLE_CUDA=1 and ENABLE_OFFLOAD=1 the case that is supposed to work. When enable CUDA is off none of my CUDA stuff is built or tested. Eventually I'd like to have abstractions such that MatrixDelay...could to be generic that only these per async contex compute classes remaining with "non portable" OMPtarget and CUDA code.

File: src/QMCWaveFunctions/Fermion/DiracMatrixComputeCUDA.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracMatrixComputeCUDA.hpp]: This is cleaned up, API just trusts the CUDALinearAlgebraHandles passed to it now.

File: src/QMCWaveFunctions/Fermion/DiracMatrixComputeCUDA.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCWaveFunctions/Fermion/DiracMatrixComputeCUDA.hpp:236]: Convinced worth possible future rework for explicit types. Changed.

**Comment by ye-luo** [comment]
Start testing in-house

**Review by ye-luo — APPROVED** [review]
Thanks!

**Pull request merged**